### PR TITLE
feat: Bandwidth/Request Metrics

### PR DESCRIPTION
**Pull request by daytime-em**
Adds request and detailed rendition/segment tracking

This doesn't include live latency, which I'm going to try to separate out into another class

### CODE REVIEW

File: library-exo/src/main/java/com/mux/stats/sdk/muxstats/ExoPlayerBinding.kt [hunk 1]
```diff
@@ -10,28 +10,57 @@ import androidx.media3.common.util.UnstableApi
...
   override fun bindPlayer(player: ExoPlayer, collector: MuxStateCollector) {
-    listener = MuxAnalyticsListener(player, collector).also { player.addAnalyticsListener(it) }
+    listener = MuxAnalyticsListener(
+      player = player,
+      collector = collector,
+      bandwidthMetrics = BandwidthMetricDispatcher(
+        player = player,
+        collector = collector,
+        trackedResponseHeaders = listOf(
+          TrackedHeader.ExactlyIgnoreCase("x-cdn"),
```
Comment (andrewjl-mux) [inline library-exo/src/main/java/com/mux/stats/sdk/muxstats/ExoPlayerBinding.kt:34]: `AVPlayer` unfortunately only lets one object listen to these at a time (without weird runtime hacks), `ExoPlayer` got this right

File: library-exo/src/main/java/com/mux/stats/sdk/muxstats/ExoPlayerBinding.kt [hunk 2]
```diff
@@ -60,19 +89,29 @@ private class MuxAnalyticsListener(player: ExoPlayer, val collector: MuxStateCol
     format: Format,
     decoderReuseEvaluation: DecoderReuseEvaluation?
   ) {
+    MuxLogger.d(
+      TAG, "onVideoInputFormatChanged: new format: bitrate ${format.bitrate}" +
+              " and frameRate ${format.frameRate} "
+    )
+    val cleanBitrate = format.bitrate.takeIf { it >= 0 } ?: 0
```
Comment (andrewjl-mux) [inline library-exo/src/main/java/com/mux/stats/sdk/muxstats/ExoPlayerBinding.kt:96]: Would this cleaning ever be done elsewhere?

File: library-exo/src/main/java/com/mux/stats/sdk/muxstats/bandwidth/BandwidthMetrics.kt [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+        player.currentTimeline
+          .getWindow(player.currentWindowIndex, currentTimelineWindow)
+      } catch (e: Exception) {
+        // Failed to obtain data, ignore, we will get it on next call
+        MuxLogger.exception(e, "BandwidthMetrics", "Failed to get current timeline")
+      }
+    }
+    val segmentData = BandwidthMetricData()
+    //RequestStart timestamp is currently not available from ExoPlayer
+    segmentData.requestResponseStart = System.currentTimeMillis()
```
Comment (andrewjl-mux) [inline library-exo/src/main/java/com/mux/stats/sdk/muxstats/bandwidth/BandwidthMetrics.kt:102]: Would it make sense to use dependency injection for `System.currentTimeMillis()`?

File: library-exo/src/main/java/com/mux/stats/sdk/muxstats/bandwidth/BandwidthMetrics.kt [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+        }
+
+        C.DATA_TYPE_MEDIA -> {
+          segmentData.requestType = "media"
+          segmentData.requestMediaDuration = (mediaEndTimeMs
+                  - mediaStartTimeMs)
+        }
+      }
+    }
+    segmentData.requestResponseHeaders = null
```
Comment (andrewjl-mux) [inline library-exo/src/main/java/com/mux/stats/sdk/muxstats/bandwidth/BandwidthMetrics.kt:134]: Would it be idiomatic Kotlin to extract the logic used to populate `segmentData` fields into a constructor?

**Review by andrewjl-mux — APPROVED** [review]

File: library-exo/src/main/java/com/mux/stats/sdk/muxstats/bandwidth/BandwidthMetrics.kt (hunk 4 above)
Comment (daytime-em) [reply library-exo/src/main/java/com/mux/stats/sdk/muxstats/bandwidth/BandwidthMetrics.kt:134]: Good callout! This kotlin came out of the auto-converter from Java, so there's plenty of opportunities to improve it

File: library-exo/src/main/java/com/mux/stats/sdk/muxstats/ExoPlayerBinding.kt (hunk 2 above)
Comment (daytime-em) [reply library-exo/src/main/java/com/mux/stats/sdk/muxstats/ExoPlayerBinding.kt:96]: As far as I know, this specific convention is just for `Format`.

File: library-exo/src/main/java/com/mux/stats/sdk/muxstats/bandwidth/BandwidthMetrics.kt (hunk 4 above)
Comment (daytime-em) [reply library-exo/src/main/java/com/mux/stats/sdk/muxstats/bandwidth/BandwidthMetrics.kt:134]: Wait no, in this case I don't think so sorry.

`SegmentData` is a core type and shouldn't have exoplayer-specific logic in it. Some other player SDK may fill in `SegmentData` differently/less completely/whatever. Kotlin extensions can't add ctors, so we just set the properties

Comment (andrewjl-mux) [reply library-exo/src/main/java/com/mux/stats/sdk/muxstats/bandwidth/BandwidthMetrics.kt:134]: > Kotlin extensions can't add ctors, so we just set the properties

Ah TIL, makes sense

**Comment by daytime-em** [comment]
I think there's some larger refactoring that can be done to promote test-ability, but I'm going to save it for after I finished ad metrics

**Pull request merged**